### PR TITLE
build xnba config files from boottarget table

### PR DESCRIPTION
**Pull request by cxhong**
for the issue from xcat-user https://sourceforge.net/p/xcat/mailman/xcat-user/thread/4E51CB2C2C427540B0CA105CCAB443C8397AAE13%40usmailmbx04/#msg36340307

Fill in kernel, initrd and kcmline from boottarget table if those values are unavailable when building xnba config files.

sample boottarget image:
````
# lsdef -t osimage c7livecd2
Object name: c7livecd2
    addkcmdline=root=live:http://11.11.101.1:/install/e8-livecd/LiveOS/squashfs.img  console=tty0 console=ttyS0,115200
    boottarget=c7livecd
    imagetype=linux
    osarch=x86_64
    osvers=boottarget
    otherpkgdir=/install/post/otherpkgs/centos7.4/x86_64
    pkgdir=/install/centos7.4/x86_64
    pkglist=/opt/xcat/share/xcat/install/centos/compute.centos7.pkglist
    profile=c7livecd
    provmethod=netboot
    template=/opt/xcat/share/xcat/install/centos/compute.centos7.tmpl
````
boottarget table:
```
]# tabdump boottarget
#bprofile,kernel,initrd,kcmdline,comments,disable
"c7livecd","vmlinuz0","initrd0.img","root=live:http://11.11.101.1:/install/e8-livecd/LiveOS/squashfs.img  console=tty0 console=ttyS0,115200",,
````
run nodeset to update xnba conf file:
```
# nodeset c910f04x42vm105 osimage=c7livecd2
c910f04x42vm105: netboot boottarget-x86_64-c7livecd

# ls -ltr /tftpboot/xcat/xnba/nodes/c910f04x42vm105*
-rw-r--r-- 1 root root 114 Jun 15 13:11 /tftpboot/xcat/xnba/nodes/c910f04x42vm105.uefi
-rw-r--r-- 1 root root 225 Jun 15 13:11 /tftpboot/xcat/xnba/nodes/c910f04x42vm105.elilo
-rw-r--r-- 1 root root 330 Jun 15 13:11 /tftpboot/xcat/xnba/nodes/c910f04x42vm105
````

### CODE REVIEW

**Comment by cxhong** [comment]
files after nodeset:
```
# cat c910f04x42vm105
#!gpxe
#netboot boottarget-x86_64-c7livecd
imgfetch -n kernel http://${next-server}/tftpboot/vmlinuz
imgload kernel
imgargs kernel console=tty0 console=ttyS0,115200 BOOTIF=01-${netX/machyp}
imgfetch http://${next-server}/tftpboot/initrd.img
imgexec kernel
# cat c910f04x42vm105.elilo
default="xCAT"
delay=0

image=/tftpboot/vmlinuz
   label="xCAT"
   initrd=/tftpboot/initrd.img
   append="console=tty0 console=ttyS0,115200 BOOTIF=%B"
# cat c910f04x42vm105.uefi
#!gpxe
chain http://${next-server}/tftpboot/xcat/elilo-x64.efi -C /tftpboot/xcat/xnba/nodes/c910f04x42vm105.elilo
````
The node able to provision.....
```
# nodestat c910f04x42vm105
c910f04x42vm105: sshd

# lsdef c910f04x42vm105
Object name: c910f04x42vm105
    arch=x86_64
    currchain=boot
    currstate=netboot boottarget-x86_64-c7livecd
    groups=vm,all
    initrd=initrd.img
    ip=10.4.42.105
    kcmdline=console=tty0 console=ttyS0,115200
    kernel=vmlinuz
    mac=42:e6:56:97:15:cb
    mgt=kvm
    netboot=xnba
    nodetype=boottarget
    os=boottarget
    postbootscripts=otherpkgs
    postscripts=syslog,remoteshell,syncfiles
    profile=c7livecd
    provmethod=c7livecd2
    serialport=0
    serialspeed=115200
    status=booted
    statustime=06-19-2018 17:02:28
    vmcpus=4
    vmhost=c910f04x42
    vmmemory=4096
    vmnics=br0
    vmstorage=dir:///var/lib/libvirt/images

```

File: xCAT-server/lib/xcat/plugins/xnba.pm [hunk 1]
```diff
@@ -171,6 +180,18 @@ sub setstate {
 
         }
     }
+
+    #fill in kernel, intrd and kcmdline from boottarget table if not available
+    unless ($kern->{kernel}) {
```
Comment (immarvin) [inline xCAT-server/lib/xcat/plugins/xnba.pm]: what is this condition `unless ($kern->{initrd})` for?

Comment (cxhong) [reply xCAT-server/lib/xcat/plugins/xnba.pm]: It seems bootparams table was empty at that.  $kern doesn't have any value returned when I was doing Unit testing.

**Comment by cxhong** [comment]
the xnba only supports x86 system, right?  i only verified this PR on x86 system. for x86 documentation, we still refer to sourceforge.  http://xcat-docs.readthedocs.io/en/stable/guides/admin-guides/manage_clusters/x86_64/index.html

**Pull request merged**